### ISSUE TITLE
[Discussion] Private Key out of range of [ 1, N - 1 ]

### ISSUE THREAD

**Issue by ricmoo** [indutny/elliptic#212]
Ahoy!

This came up recently in an [ethers.js issue](https://github.com/ethers-io/ethers.js/issues/702), while the original question ended up being a misunderstanding, raised another question...

What *should* happen if the private key is greater than N. Currently the library goes to work as if everything is fine (which in many ways it is). The private key of `1` creates the same public key (and ilk) as the private key `N + 1`, and everything is just peachy.

The question is, should it? And this might be a decision for consuming libraries to to make rather than elliptic itself, but I just wanted to get someone (i.e. @indutny) with more knowledge and understanding of the universe of ECC to convince me one way or the other.

On the one hand, as it stands today, a completely random string of 32 bytes will generate a private key, with only 2 possible invalid keys (i.e. 0 and N), for a `2 in 2^256` or `1 in 2^255` chance of things collapsing into confusion, which has pretty slim odds.

Other the other hand, this means that all keys in the open interval ( N, 2^256 ) suffer from modulo bias into the closed interval [ 1, 2^256 - N - 1 ], so those `2^256 - N` (ish) keys are "less secure".

So, the question is should ethers validate the range is in [ 1, N - 1 ]? Should elliptic? Should anyone?

Also, in the current methods, if either 0 or N is used, a non-obvious error (i.e. a null showing up in the RED algorithms of BN.js) is emitted. I will probably add logic to ethers.js to emit an "invalid key" error or something semi-informational, unless you feel like adding a similar check and error to elliptic directly?

Basically, elliptic is awesome, and my life is so much easier because of it. I don't mind making a PR, but I also don't mind fixing my library downstream if it makes more sense to do it there.

Thanks!

**Note:** *these examples were all assuming secp256k1, which is the only curve I've tried this on*

**Comment by ricmoo** [comment]
So, I've spent the last while reviewing other secp256k1 libraries, and it doesn't seem like any other popular libraries I've found mitigate the inherent modulo bias.

The most important thing is that 0 and N both fail to generate private keys, this is a (1 / (2 ** 255)) chance.

The modulo bias is actually quite small (((2 ** 256) - N) / (2 ** 256)), or 0.0000000000000000000000000000000000000037% (give or take a few zeros, this is largely by hand as normal REPL environments cannot handle these numbers).

So, for now, I'm going to stick with letting the underlying library do what it does, and in this case elliptic folds them (implicitly) into their (x mod N) equivalent. But if anyone comes across this, or anyone has counter reasons, please feel free to re-open this issue.

Thanks! :)

**Comment by ricmoo** [comment]
(sorry I meant to comment on my own issue, not this library's issue)

I'll leave it closed though, and anyone who wishes to re-open it may.
